CSharpMath/Structures/Trie.cs

```diff
@@ -0,0 +1,216 @@
+      out ReadOnlyMemory<T> commonHead, out ReadOnlyMemory<T> thisRest, out ReadOnlyMemory<T> otherRest) {
+      var thisSpan = @this.Span;
+      var otherSpan = other.Span;
+      var splitIndex = 0;
+      while (
+        splitIndex < thisSpan.Length
+        && splitIndex < otherSpan.Length
+        && otherSpan[splitIndex] is var o
+        && (thisSpan[splitIndex]?.Equals(o) ?? o is null)
+      ) splitIndex++;
+      commonHead = @this.Slice(0, splitIndex);
+      thisRest = @this.Slice(splitIndex);
+      otherRest = other.Slice(splitIndex);
+    }
+    public static void ZipWith<T>(this ReadOnlySpan<T> @this, ReadOnlySpan<T> other,
+      out ReadOnlySpan<T> commonHead, out ReadOnlySpan<T> thisRest, out ReadOnlySpan<T> otherRest) {
+      var splitIndex = 0;
+      while (
+        splitIndex < @this.Length
+        && splitIndex < other.Length
+        && other[splitIndex] is var o
+        && (@this[splitIndex]?.Equals(o) ?? o is null)
+      ) splitIndex++;
+      commonHead = @this.Slice(0, splitIndex);
+      thisRest = @this.Slice(splitIndex);
+      otherRest = other.Slice(splitIndex);
+    }
+  }
+}
+
+namespace CSharpMath.Structures {
+  // Based on https://github.com/gmamaladze/trienet/tree/f0cce5f980d85e445188b3eb025821fcdb740144/TrieNet/_PatriciaTrie
+  // Can't use the TrieNet NuGet package because the .NET Standard 2.0 version is not uploaded: https://github.com/gmamaladze/trienet/issues/12
+  [Serializable]
+  [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix",
+    Justification = "'Trie' is the correct data type, not 'Collection'")]
+  public class PatriciaTrie<TKeyElement, TValue> : IEnumerable<KeyValuePair<ReadOnlyMemory<TKeyElement>, TValue>> {
+    #region Originally TrieNodeBase<TValue>
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1043:Use Integral Or String Argument For Indexers",
+      Justification = "ReadOnlySpan<char> is basically string but faster")]
+    public IEnumerable<TValue> this[ReadOnlySpan<TKeyElement> query] =>
+      query.IsEmpty ? ValuesDeep() : SearchDeep(query);
+    protected IEnumerable<TValue> SearchDeep(ReadOnlySpan<TKeyElement> query) =>
+      GetChildOrNull(query) is { } nextNode
+      ? nextNode[query.Slice(Math.Min(query.Length, nextNode.Key.Length))]
+      : Enumerable.Empty<TValue>();
+    private IEnumerable<TValue> ValuesDeep() => Subtree().SelectMany(node => node.Values);
+    protected IEnumerable<PatriciaTrie<TKeyElement, TValue>> Subtree() =>
+      Enumerable.Repeat(this, 1).Concat(Children.Values.SelectMany(child => child.Subtree()));
+    #endregion Originally TrieNodeBase<TValue>
+
+    protected Dictionary<TKeyElement, PatriciaTrie<TKeyElement, TValue>> Children { get; private set; }
+    protected ReadOnlyMemory<TKeyElement> Key { get; private set; }
+    protected Queue<TValue> Values { get; private set; }
+
+    public PatriciaTrie() : this(
```

Comment:
The question is what is the purpose of using a Trie here. Moving conversation to here:

> Tries are used to lookup values with the same key prefix, compared to a Dictionary which requires a complete key match. This enables us to e.g. convert multi-char symbols (esp. for characters above the Unicode BMP) like 𝐀 (U+1D400, represented in UTF-16 as U+D835 U+DC00) to a new Variable("A") { FontStyle = FontStyle.Bold }.

Searching on strings is fast. Also the example you give is of marginal importance because there is no need to use unicode "Mathematical Bold" ranges except at the point of rendering of LaTeX fonts. The lookup doesn't need to go in this direction therefore: the direction is, in special situations (when there is no standard italic font and the TeX font is being relied on), to go from an object to the unicode mathematical bold. If the need for the Trie is to support font styles please wait for my PR which will standardize Bold/Italic fonts.

We need some estimates on the benefits of the Trie. Presumably these are performance-related?

Reply:
The point is that we should convert `𝐀` to a `Variable` correctly instead of defaulting to `Ordinary`. This is similar to supporting e.g. converting `ů` to be `new Accent("\u030A", new Variable("u"))` as seen in https://github.com/verybadcat/CSharpMath/issues/136#issuecomment-647143383 so that font style changes can be supported correctly.
>Also the example you give is of marginal importance

This is just opinion.
>If the need for the Trie is to support font styles please wait for my PR which will standardize Bold/Italic fonts.

This is just one use case. By using a trie, we can support parsing arbitrary-length sequences of `char`s as a single group.

Reply:
> converting ů to be new Accent("\u030A", new Variable("u"))

That's another marginal use case as you can also define these using latex accent syntax, and it's extremely rare to have this accent as a mathematical variable. It's rare enough to have bold characters as mathematical variables, and when you do these will almost never be specified using the "Mathematical bold" unicode range. They will be specified using latex bold syntax.

Regardless of where this might possibly be used, if it is used in such cases, what is the benefit? Performance? I have asked this twice before.

The cost of adding this feature is large and I am pretty sure the benefit is negligible.

Reply:
>It's marginal because it's extremely rare to have characters use this unicode range (outside of the internals of latex libraries)

This is for correctness of Unicode rendering. We should leave a way for users to define arbitrary characters as `BinaryOperator`, `LargeOperator` etc. You could write conditionals inside handlers for single surrogates but you would be writing a trie in code instead of using an existing structure to help.
>Font support will deal with this [...] It's just an analysis of a char...

The sentence started with a "This is similar to supporting e.g. " to provide an example where a character itself is converted to an atom with more complex structure: We should support this for all characters (not just whole code points, consider the existence of Variation Selectors etc.). If a user wants `ff` to be parsed as `ﬀ` (Latin Small Ligature Ff), this can be done too.

More examples can be seen in https://github.com/verybadcat/CSharpMath/pull/143/files#diff-117f6680b81ca124968fcd26ad2fe561R893-R896 where one to four primes are parsed as one character to enable better render results and in the future https://tex.stackexchange.com/a/53637 where one to three dashes are parsed as one character.
>What is the benefit of this compared to just matching strings?

How would you "match" a string? As mentioned above, `Tries are used to lookup values with the same key prefix, compared to a Dictionary which requires a complete key match.` We use the right structures for the right scenarios and we don't spam `Dictionary` everywhere.

Reply:
I don't see why the parser is working one character at a time. I understand there are particular characters (e.g. `\`) that should be looked for (in that case, starting a search for an appropriate command structure).

Suppose that "tt" means something special for the sake of argument. Then you can have a "subparser" that looks for "tt" at the start of the string. You don't need a Trie to do that. You only need String.StartsWith.

Reply:
>I don't see why the parser is working one character at a time. I understand there are particular characters (e.g. \) that should be looked for (in that case, starting a search for an appropriate command structure).

How else would you enumerate a string? We don't have `System.Text.Rune` on .NET Standard.
>Suppose that "tt" means something special for the sake of argument. Then you can have a "subparser" that looks for "tt" at the start of the string. You don't need a Trie to do that. You only need String.StartsWith.

You mean instead of using a Dictionary with a near-constant-time lookup as is implemented in the current trie, we do a linear search instead? Why should we throw away existing data structures and replace them with slower loops?

If you want to throw away data structures, why are we even using `Dictionary` at all? We can just do a very large loop on an array each time we want to look up a command.

Reply:
And remember that lookups are performed on each `char` in a string, so the lookup time will accumulate over large LaTeX.